packages/core-contracts/contracts/utils/ERC20.sol

```diff
@@ -0,0 +1,93 @@
+//SPDX-License-Identifier: UNLICENSED
+
+pragma solidity ^0.8.0;
+
+import "../interfaces/IERC20.sol";
+
+abstract contract ERC20 is IERC20 {
+    string public override name;
+
+    string public override symbol;
+
+    uint8 public immutable override decimals;
+
+    mapping(address => uint256) public override balanceOf;
+
+    mapping(address => mapping(address => uint256)) public override allowance;
+
+    uint256 public override totalSupply;
+
+    constructor(
+        string memory tokenName,
+        string memory tokenSymbol,
+        uint8 tokenDecimals
+    ) {
+        name = tokenName;
+        symbol = tokenSymbol;
+        decimals = tokenDecimals;
+    }
+
+    function approve(address spender, uint256 amount) public override returns (bool) {
+        allowance[msg.sender][spender] = amount;
+
+        emit Approval(msg.sender, spender, amount);
+
+        return true;
+    }
+
+    function transfer(address to, uint256 amount) public override returns (bool) {
+        _transfer(msg.sender, to, amount);
+
+        return true;
+    }
+
+    function transferFrom(
+        address from,
+        address to,
+        uint amount
+    ) external override returns (bool) {
+        // it reverts if the sender does not have sufficient allowance
+        allowance[from][msg.sender] -= amount;
```

Comment:
What if `from = msg.sender`? Is it ERC20 compatible to revert in those cases?

Reply:
Also, this would produce an error that is not very explicit.

Reply:
It is erc20 compatible. The other approach will be to have explicit requirements e.g.  `require(allowance >= amount)`
and then do the unchecked operation. But in my opinion an underflow error there is quite informative and it is whatyou get in most cases when using safe math.